### PR TITLE
feat(server): allow handle labeling rules for datasets from API

### PR DESCRIPTION
**Pull request by frascuchon**
This PR includes dataset labeling rules management for API.

- Following the API rule data model:

LabelingRule | Data model for labeling rules creation
-- | --
query* | The rule es query
label* | The linked rule label
description | Description related to rule

- Rules are keyed by query, we cannot include rules with same query

@dcfidalgo @dvsrepo  Feedback welcome

See #745

### CODE REVIEW

**Review by dvsrepo — APPROVED** [review]
Niiiice. My only questions would be:

- would this accommodate in the future for multilabel rules?
- for token classification besides from the query and the label we will need additional information (if the label should be applied to the matching token, or to the token in the next i positions, etc), just to keep it mind and if something needs to be taken into account in this first design.
- ideally, I'd like to have the agent adding the rule (annotation agent should do it) in the same way we do for annotations in the UI if it's not too much trouble, I foresee rules can be defined much more collaboratively and this info would be useful.
That's all, I like the design a lot!

File: src/rubrix/server/tasks/text_classification/service/service.py [hunk 1]
```diff
@@ -168,29 +203,52 @@ def _check_multi_label_integrity(
...
+            The rule
+
+        """
+        self.__labeling__.add_rule(dataset, rule)
+
+    def delete_labeling_rule(self, dataset: Dataset, rule_query: str):
+        """
+        Deletes a rule from a dataset.
+
+        Nothing happens if rule does exists in dataset.
```
Comment (dcfidalgo) [inline src/rubrix/server/tasks/text_classification/service/service.py]:
```suggestion
        Nothing happens if the rule does not exist in dataset.
```

File: src/rubrix/server/datasets/dao.py [hunk 2]
```diff
@@ -51,9 +52,35 @@ def dataset_records_index(dataset_id: str) -> str:
...
     """Datasets DAO"""
 
+    _INSTANCE = None
+
+    @classmethod
+    def get_instance(
+        cls, es: ElasticsearchWrapper = Depends(create_es_wrapper)
+    ) -> "DatasetsDAO":
+        """
+        Get or creates the dao instance
```
Comment (dcfidalgo) [inline src/rubrix/server/datasets/dao.py]:
```suggestion
        Gets or creates the dao instance
```

File: src/rubrix/server/datasets/dao.py [hunk 3]
```diff
@@ -51,9 +52,35 @@ def dataset_records_index(dataset_id: str) -> str:
...
+
+        Parameters
+        ----------
+        es:
+            The elasticsearch wrapper dependency>
```
Comment (dcfidalgo) [inline src/rubrix/server/datasets/dao.py]:
```suggestion
            The elasticsearch wrapper dependency
```

File: src/rubrix/server/tasks/text_classification/api/model.py [hunk 4]
```diff
@@ -35,6 +35,67 @@
...
 
+class CreateLabelingRule(BaseModel):
+    """
+    Data model for labeling rules creation
+
+    Attributes:
+    -----------
+
+    query:
+        The rule es query
```
Comment (dcfidalgo) [inline src/rubrix/server/tasks/text_classification/api/model.py]:
```suggestion
        The ES query of the rule
```

File: src/rubrix/server/tasks/text_classification/api/model.py [hunk 5]
```diff
@@ -35,6 +35,67 @@
...
+
+    label: str
+        The linked rule label
```
Comment (dcfidalgo) [inline src/rubrix/server/tasks/text_classification/api/model.py]:
```suggestion
        The label associated with the rule
```

File: src/rubrix/server/tasks/text_classification/api/model.py [hunk 6]
```diff
@@ -35,6 +35,67 @@
...
+
+    description:
+        Description related to rule
```
Comment (dcfidalgo) [inline src/rubrix/server/tasks/text_classification/api/model.py]:
```suggestion
        A brief description of the rule
```

File: src/rubrix/server/tasks/text_classification/api/model.py [hunk 7]
```diff
@@ -35,6 +35,67 @@
...
+
+    @validator("query")
+    def strip_query(cls, query: str) -> str:
+        """Remove blank spaces for query"""
+        return query.strip()
+
+
+class LabelingRule(CreateLabelingRule):
+    """
+    Adds read-only attributes to labeling rule
```
Comment (dcfidalgo) [inline src/rubrix/server/tasks/text_classification/api/model.py]:
```suggestion
    Adds read-only attributes to the labeling rule
```

File: src/rubrix/server/tasks/text_classification/api/model.py [hunk 8]
```diff
@@ -35,6 +35,67 @@
...
+
+    Attributes:
+    -----------
+
+    author:
+        Who has created the rule
+
+    created_at:
+        When rule was created
```
Comment (dcfidalgo) [inline src/rubrix/server/tasks/text_classification/api/model.py]:
```suggestion
        When was the rule created
```

File: src/rubrix/server/tasks/text_classification/api/model.py [hunk 9]
```diff
@@ -35,6 +35,67 @@
...
+
+class LabelingRule(CreateLabelingRule):
+    """
+    Adds read-only attributes to labeling rule
+
+    Attributes:
+    -----------
+
+    author:
+        Who has created the rule
```
Comment (dcfidalgo) [inline src/rubrix/server/tasks/text_classification/api/model.py]:
```suggestion
        Who created the rule
```

File: src/rubrix/server/tasks/text_classification/service/service.py [hunk 10]
```diff
@@ -168,29 +203,52 @@ def _check_multi_label_integrity(
...
 
-_instance = None
+        Parameters
+        ----------
+        dataset:
+            The dataset
 
+        Returns
+        -------
+            A list of labeling rules create for a given dataset
```
Comment (dcfidalgo) [inline src/rubrix/server/tasks/text_classification/service/service.py]:
```suggestion
            A list of labeling rules for a given dataset
```

**Review by dcfidalgo — APPROVED** [review]
great, really nice!! Just a few minor inline suggestions. Two questions though, is the `NEW_BASE_ENDPOINT` supposed to replace the `BASE_ENDPOINT` in the long term? The second one, in the client code we try to follow google style docstrings, since they work better together with sphinx when taking type annotations into account. should we try to follow the same style in the server code?

**Comment by frascuchon** [comment]
> great, really nice!! Just a few minor inline suggestions. Two questions though, is the `NEW_BASE_ENDPOINT` supposed to replace the `BASE_ENDPOINT` in the long term? The second one, in the client code we try to follow google style docstrings, since they work better together with sphinx when taking type annotations into account. should we try to follow the same style in the server code?

For the first question, YES. I want to move the api path in the form `/api/datasets/{task}/{name}/...`. The current format was adopted when a dataset could wrap several task records. So, I prefer create new endpoints with this new format. 

For the second I will create a tech debt issue planning docstring migration to google style.

**Comment by frascuchon** [comment]
> Niiiice. My only questions would be:
> 
> * would this accommodate in the future for multilabel rules?

Only text classification with single label are supported as weak supervision paradigm in rubrix. Do we known all we need for design data model for multi-label weak supervision? We can define `label` as  `Union[str, List[str]]` but not sure if we're still missing something else. 

> * for token classification besides from the query and the label we will need additional information (if the label should be applied to the matching token, or to the token in the next i positions, etc), just to keep it mind and if something needs to be taken into account in this first design.

The rule data model only affects to text classification task. The token classification task will have it own labeling rule data model.

> * ideally, I'd like to have the agent adding the rule (annotation agent should do it) in the same way we do for annotations in the UI if it's not too much trouble, I foresee rules can be defined much more collaboratively and this info would be useful.
>   That's all, I like the design a lot!

Described fields are input or modifiable fields. There are also 2 read-only fields for the [API response data model](https://github.com/recognai/rubrix/blob/cfe1cd4bcc1a7d51f6bd7d81e2c3b2f3da762890/src/rubrix/server/tasks/text_classification/api/model.py#L66): `author` and `created_at`

```python
class LabelingRule(CreateLabelingRule):
    """
    Adds read-only attributes to the labeling rule
    Attributes:
    -----------
    author:
        Who created the rule
    created_at:
        When was the rule created
    """

    author: str
    created_at: Optional[datetime] = Field(default_factory=datetime.utcnow)
```

The rule author will be automatically filled with username who makes the creation request. We can normalizes the name and use `agent` but I think `author` match better for rule creation.

**Comment by dvsrepo** [comment]
> > Niiiice. My only questions would be:
> > 
> > * would this accommodate in the future for multilabel rules?
> 
> Only text classification with single label are supported as weak supervision paradigm in rubrix. Do we known all we need for design data model for multi-label weak supervision? We can define `label` as `Union[str, List[str]]` but not sure if we're still missing something else.


Yes, what I mean is that we plan to support it in the near future just want to make sure there are no big issues when we do so, but we need to define how the weak label matrix works, etc. As we did for single label.
> 
> > * for token classification besides from the query and the label we will need additional information (if the label should be applied to the matching token, or to the token in the next i positions, etc), just to keep it mind and if something needs to be taken into account in this first design.
> 
> The rule data model only affects to text classification task. The token classification task will have it own labeling rule data model.


Ok, good
> 
> > * ideally, I'd like to have the agent adding the rule (annotation agent should do it) in the same way we do for annotations in the UI if it's not too much trouble, I foresee rules can be defined much more collaboratively and this info would be useful.
> >   That's all, I like the design a lot!
> 
> Described fields are input or modifiable fields. There are also 2 read-only fields for the [API response data model](https://github.com/recognai/rubrix/blob/cfe1cd4bcc1a7d51f6bd7d81e2c3b2f3da762890/src/rubrix/server/tasks/text_classification/api/model.py#L66): `author` and `created_at`
> 
> ```python
> class LabelingRule(CreateLabelingRule):
>     """
>     Adds read-only attributes to the labeling rule
>     Attributes:
>     -----------
>     author:
>         Who created the rule
>     created_at:
>         When was the rule created
>     """
> 
>     author: str
>     created_at: Optional[datetime] = Field(default_factory=datetime.utcnow)
> ```
> 
> The rule author will be automatically filled with username who makes the creation request. We can normalizes the name and use `agent` but I think `author` match better for rule creation.

Good, I hadn't seen this. Not strong opinion on the naming, although I always prefer to have more aligned names especially when they work so similarly (in the sense that is the username performing the action: annotating or adding a rule).

Great job!!

**Pull request merged**